iree/hal/cts/command_queue_test.cc

```diff
@@ -0,0 +1,140 @@
+  // Log how many queues we have so future test cases have more context.
+  // Most tests just use the first queue, but supporting multiple queues may be
+  // relevant on some implementations.
+
+  absl::Span<CommandQueue*> dispatch_queues = device_->dispatch_queues();
+  LOG(INFO) << "Device has " << dispatch_queues.size() << " dispatch queue(s)";
+  EXPECT_GE(dispatch_queues.size(), 1);
+  for (auto* dispatch_queue : dispatch_queues) {
+    EXPECT_TRUE(dispatch_queue->can_dispatch());
+  }
+
+  absl::Span<CommandQueue*> transfer_queues = device_->transfer_queues();
+  LOG(INFO) << "Device has " << transfer_queues.size() << " transfer queue(s)";
+  EXPECT_GE(transfer_queues.size(), 1);
+  for (auto* transfer_queue : transfer_queues) {
+    EXPECT_TRUE(transfer_queue->can_transfer());
+  }
+}
+
+// Tests that waiting for idle is a no-op when nothing is queued.
+TEST_P(CommandQueueTest, WaitIdleWhileIdle) {
+  for (auto* dispatch_queue : device_->dispatch_queues()) {
+    EXPECT_OK(dispatch_queue->WaitIdle());
+  }
+  for (auto* transfer_queue : device_->transfer_queues()) {
+    EXPECT_OK(transfer_queue->WaitIdle());
+  }
+}
+
+// Tests that submitting a command buffer and immediately waiting will not
+// deadlock.
+TEST_P(CommandQueueTest, BlockingSubmit) {
+  auto command_queue = device_->dispatch_queues()[0];
+
+  ASSERT_OK_AND_ASSIGN(auto command_buffer, device_->CreateCommandBuffer(
+                                                CommandBufferMode::kOneShot,
+                                                CommandCategory::kDispatch));
+  ASSERT_OK_AND_ASSIGN(auto semaphore, device_->CreateSemaphore(0ull));
+
+  ASSERT_OK(command_queue->Submit(
+      {{}, {command_buffer.get()}, {{semaphore.get(), 1ull}}}));
+  ASSERT_OK(semaphore->Wait(1ull, absl::InfiniteFuture()));
+}
+
+// Tests waiting while work is pending/in-flight.
+TEST_P(CommandQueueTest, WaitTimeout) {
+  auto command_queue = device_->dispatch_queues()[0];
+
+  ASSERT_OK_AND_ASSIGN(auto command_buffer, device_->CreateCommandBuffer(
+                                                CommandBufferMode::kOneShot,
+                                                CommandCategory::kDispatch));
+  ASSERT_OK_AND_ASSIGN(auto wait_semaphore, device_->CreateSemaphore(0ull));
+  ASSERT_OK_AND_ASSIGN(auto signal_semaphore, device_->CreateSemaphore(0ull));
+
+  ASSERT_OK(command_queue->Submit({{{wait_semaphore.get(), 1ull}},
+                                   {command_buffer.get()},
+                                   {{signal_semaphore.get(), 1ull}}}));
+
+  // Work shouldn't start until the wait semaphore reaches its payload value.
+  EXPECT_THAT(signal_semaphore->Query(), IsOkAndHolds(Eq(0ull)));
+  EXPECT_TRUE(
+      IsDeadlineExceeded(command_queue->WaitIdle(absl::Milliseconds(100))));
+
+  // Signal the wait semaphore, work should begin and complete.
+  ASSERT_OK(wait_semaphore->Signal(1ull));
+  ASSERT_OK(signal_semaphore->Wait(1ull, absl::InfiniteFuture()));
```

Comment:
With emulated timeline semaphores (introduced in https://github.com/google/iree/pull/2208), this never completes